fix: Make it explicit that there is a network call to MDS to get SecureSessionAgentConfig

rmehta19 opened the pull request:
Make the classes non-final while we're here to allow for mocking in tests if needed.

lqiu96 commented on oauth2_http/java/com/google/auth/oauth2/S2A.java:
```diff
@@ -58,7 +58,7 @@
  */
 @ThreadSafe
-public final class S2A {
+public class S2A {
```
On a second look, do you have concerns if we can make this name written out specifically (SecureSessionAgent.java) and (SecureSessionAgentConfig.java) as S2A may not be well known and specific to Google?

rmehta19 replied:
That's fine with me, done in https://github.com/googleapis/google-auth-library-java/pull/1573/commits/e6b6e18582b7538a2ea6082c13885355016fbcd4

lqiu96 commented on oauth2_http/java/com/google/auth/oauth2/S2A.java:
```diff
@@ -81,14 +81,9 @@ public final class S2A {
-  /** @return the plaintext S2A Address from the mTLS config. */
-  public String getPlaintextS2AAddress() {
-    return config.getPlaintextAddress();
+  /** @return the cached S2AConfig. */
+  public S2AConfig getConfigFromMDS() {
+    return config;
   }
```
What I had in mind was
```
public SecureSessionAgentConfig getConfig() {
  return getS2AConfigFromMDS();
}
```
as this `getConfig()` call is the explicit call that will make the network call.

rmehta19 replied:
I see, so right now the config is being set in the builder and is a member of the class. The current setup ensures that only one call to MDS is made (since there is no need to make the request to MDS multiple times) no matter how many times a user calls getConfigFromMds.

I'm fine with this though, it shifts responsibility to the user to store config and not call getConfig() more than once to save latency of their application. This seems aligned with what you mentioned in our offline discussion yesterday about how it should be explicit a network call is being made.

Done in: https://github.com/googleapis/google-auth-library-java/pull/1573/commits/381efe4531e481a2f4a36fbaa70dc94ce55ecc38

lqiu96 commented on oauth2_http/java/com/google/auth/oauth2/S2A.java:
```diff
@@ -81,14 +81,9 @@ public final class S2A {
+    return config;
   }
 
   public static Builder newBuilder() {
```
Also, if you want, you can include a static `create()` helper that creates a default instance (i.e. `SecureSessionAgent.create()`).

rmehta19 replied:
Done in https://github.com/googleapis/google-auth-library-java/pull/1573/commits/7ce029110e5752ae469e46174c97e30d65caa29c

lqiu96 commented on oauth2_http/java/com/google/auth/oauth2/SecureSessionAgent.java:
```diff
@@ -58,7 +58,7 @@
  */
 @ThreadSafe
-public final class S2A {
+public class SecureSessionAgent {
```
Can we keep final or does that break something?

lqiu96 replied:
Can you point me to where we need to mock it? I think we should be able to find other ways around, but just want to check.

rmehta19 replied:
In https://github.com/googleapis/sdk-platform-java/pull/3326 if we want to test: `createS2ASecuredChannelCredentials` function added. I was thinking Mockito would be the simplest way to mock this lib, in which case I believe the class would have to be non-final

lqiu96 replied:
Do you have an example of how we're testing in the PR right now?

One option that comes to mind for testing this without mocks:
Update the method to: `createS2ASecuredChannelCredentials(SecureSessionAgent s2a)`

The tests for createS2ASecuredChannelCredentials would pass version of SecureSessionAgent (i.e. SecureSessionAgentTest which overrides the `getConfig()` and returns SecureSessionAgentConfig with the needed values). 

I *think* it's possible to test it without having to mock it.

lqiu96 replied:
Ah ignore the above, I was mistaken. I am fine with removing the final keyword.

lqiu96 commented on oauth2_http/java/com/google/auth/oauth2/SecureSessionAgent.java:
```diff
@@ -72,23 +72,24 @@ public final class S2A {
-  /** @return the plaintext S2A Address from the mTLS config. */
-  public String getPlaintextS2AAddress() {
-    return config.getPlaintextAddress();
+  public static SecureSessionAgent create() {
+    return newBuilder().build();
```
```suggestion
  /** @return default instance of SecureSessionAgent */
  public static SecureSessionAgent create() {
    return newBuilder().build();
```

or something similar

lqiu96 commented on oauth2_http/java/com/google/auth/oauth2/SecureSessionAgentConfig.java:
```diff
@@ -33,7 +33,7 @@
 
 /** Holds an mTLS configuration (consists of address of S2A) retrieved from the Metadata Server. */
-final class S2AConfig {
+public class SecureSessionAgentConfig {
```
Can this continue to be final?

rmehta19 replied:
final will let us mock in tests (if that's something we want to test in sdk PR). If not, I'm fine keeping this final

lqiu96 reviewed (approved):
LGTM. Added a few nits/ questions

lqiu96 commented:
@rmehta19  Also can you update the title with `fix: {title}`?

rmehta19 commented:
@lqiu96 , Thanks for the review! Addressed comments.

lqiu96 commented:
> @lqiu96 , Thanks for the review! Addressed comments.

LGTM. Let me know if you feel strongly about final/ non-final. I'll approve again and we can merge then

lqiu96 commented:
FYI the linter is failing. I think you need to run `mvn fmt:format` again

rmehta19 commented:
> > @lqiu96 , Thanks for the review! Addressed comments.
> 
> LGTM. Let me know if you feel strongly about final/ non-final. I'll approve again and we can merge then

As discussed offline, let's keep it final so that we can test it from Gax where it is used
